packages/auth/src/grant/routes.test.ts

```diff
@@ -858,70 +811,260 @@ describe('Grant Routes', (): void => {
         ).resolves.toBeUndefined()
         expect(ctx.status).toBe(302)
         expect(redirectSpy).toHaveBeenCalledWith(clientRedirectUri.toString())
+      })
+
+      test('Can finish rejected interaction', async (): Promise<void> => {
+        const grant = await Grant.query().insert({
+          ...generateBaseGrant(),
+          state: GrantState.Rejected
+        })
+
+        await Access.query().insert({
+          ...BASE_GRANT_ACCESS,
+          grantId: grant.id
+        })
+
+        const ctx = createContext(
+          {
+            headers: {
+              Accept: 'application/json',
+              'Content-Type': 'application/json'
+            },
+            session: {
+              nonce: grant.interactNonce
+            }
+          },
+          { id: grant.interactId, nonce: grant.interactNonce }
+        )
+
+        const clientRedirectUri = new URL(grant.finishUri)
+        clientRedirectUri.searchParams.set('result', 'grant_rejected')
+
+        const redirectSpy = jest.spyOn(ctx, 'redirect')
+
+        await expect(
+          grantRoutes.interaction.finish(ctx)
+        ).resolves.toBeUndefined()
+        expect(ctx.status).toBe(302)
+        expect(redirectSpy).toHaveBeenCalledWith(clientRedirectUri.toString())
+      })
+
+      test('Cannot finish invalid interaction', async (): Promise<void> => {
+        const ctx = createContext(
+          {
+            headers: {
+              Accept: 'application/json',
+              'Content-Type': 'application/json'
+            },
+            session: {
+              nonce: grant.interactNonce
+            }
+          },
+          { id: grant.interactId, nonce: grant.interactNonce }
+        )
+
+        const clientRedirectUri = new URL(grant.finishUri)
+        clientRedirectUri.searchParams.set('result', 'grant_invalid')
+
+        const redirectSpy = jest.spyOn(ctx, 'redirect')
+
+        await expect(
+          grantRoutes.interaction.finish(ctx)
+        ).resolves.toBeUndefined()
+        expect(ctx.status).toBe(302)
+        expect(redirectSpy).toHaveBeenCalledWith(clientRedirectUri.toString())
+      })
+    })
+  })
 
-        const issuedGrant = await Grant.query().findById(grant.id)
-        expect(issuedGrant.state).toEqual(GrantState.Granted)
+  describe('accept grant', (): void => {
+    test('cannot accept grant without secret', async (): Promise<void> => {
+      const ctx = createContext(
+        {
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json'
+          }
+        },
+        { id: grant.interactId }
```

Comment:
Should the `nonce` be included to be sure the `invalid_interaction` is due to the missing secret?